src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp

```diff
@@ -0,0 +1,236 @@
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of any author or any participating institution
+//    may be used to endorse or promote products derived from this software
+//    without specific prior written permission.
+// For a full list of authors, refer to the file AUTHORS.
+// --------------------------------------------------------------------------
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Lars Nilse $
+// $Authors: Lars Nilse $
+// --------------------------------------------------------------------------
+
+#include <OpenMS/KERNEL/StandardTypes.h>
+#include <OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h>
+
+#include <vector>
+#include <algorithm>
+
+using namespace std;
+
+namespace OpenMS
+{
+  HashedSpectrum::HashedSpectrum(MSSpectrum<Peak1D>& raw_spectrum, double mz_bin, double mz_tolerance, bool mz_unit_ppm)
+  : mz_bin_(mz_bin), mz_tolerance_(mz_tolerance), mz_unit_ppm_(mz_unit_ppm)
+  {
+    if (raw_spectrum.size() <= 2)
+    {
+      throw Exception::InvalidSize(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, raw_spectrum.size());
+    }
+ 
+    mz_min_ = raw_spectrum.begin()->getMZ();
+      
+    std::vector<double> mz;
+    std::vector<double> intensity;
+    int last_index = 0;
+    HashedSpectrum::MzInterval bin;
+    bin.first = &(*raw_spectrum.begin());
+    bin.last = &(*raw_spectrum.begin());
+    for (MSSpectrum<Peak1D>::Iterator it = raw_spectrum.begin(); it != raw_spectrum.end(); ++it)
+    {  
+      int index = getIndex_(it->getMZ());
+  
+      if (index > last_index)
+      {
+        // close previous interval
+        bins_.push_back(bin);
+   
+        // add empty interval
+        bin.first = NULL;
+        bin.last = NULL;
+        for (int i=0; i < (index - last_index -1); ++i)
```

Comment:
Adds bins that do not contain any data points. In this case both pointer (soon iterators) are `NULL`.